Closes a few issues

r-salas opened the pull request:
Add more info to log, add platform details, reorganize var and preprocessors directories, merge preprocessors and transformers, fix MTE API, remove -c argument when executing from docker, option to read JSON for Kaggle credentials

Commit:
Added documentation for gymnos-in-depth section: architecture, execution results and tracking

Commit:
Force Docker execution with bash (closes #69 )

Commit:
Force Docker execution with bash (closes #69 )

Commit:
Add platform details to metrics

Commit:
Add platform details to metrics

Commit:
Fix Kaggle credentials with JSON file (closes #75 )

Commit:
Fix Kaggle credentials with JSON file (closes #75 )

Commit:
Fix Movistar API url for MTE

Commit:
Fix Movistar API url for MTE

Commit:
Simplify preprocessors and transformers (closes #81)

Commit:
Simplify preprocessors and transformers (closes #81)

Commit:
Reorganize keras var directory

Commit:
Reorganize keras var directory

Commit:
Add more keras layers

Commit:
Add more keras layers

Commit:
Fix PEP8 E302

Commit:
Fix PEP8 E302

Commit:
Add to log the names of applied preprocessors (closes #67 )

Commit:
Add to log the names of applied preprocessors (closes #67 )

Commit:
Add to log a summary of the train metrics (closes #68)

Commit:
Add to log a summary of the train metrics (closes #68)

Commit:
Add some platform details to log (closes #64)

Commit:
Add some platform details to log (closes #64)

Commit:
Reorganize preprocessors to folders

Commit:
Reorganize preprocessors to folders

Commit:
Show tensorflow logs with docker

Commit:
Show tensorflow logs with docker

Commit:
Add GPU and CPU info to metrics

Commit:
Add GPU and CPU info to metrics

Commit:
Make experiment field as optional field

Commit:
Make experiment field as optional field

Commit:
Simplify trackers adding start method

Commit:
Simplify trackers adding start method

Commit:
Input shape as model parameter

Commit:
Input shape as model parameter

Commit:
Remove unused methods and variables

Commit:
Remove unused methods and variables

Commit:
Fix possible confusion implementing read method for libraryDataset (closes #82)

Commit:
Fix possible confusion implementing read method for libraryDataset (closes #82)

Commit:
Merge pull request #84 from Telefonica/documentation

Added documentation for gymnos-in-depth section: architecture, execut…

Commit:
Simplify model API and add model mixins (closes #73)

Commit:
Simplify model API and add model mixins (closes #73)

Commit:
Add tensorflow mixin

Commit:
Add tensorflow mixin

Commit:
Remove unused variable

Commit:
Remove unused variable

Commit:
Add session config example for tf model

Commit:
Add session config example for tf model

Commit:
Remove old variable

Commit:
Remove old variable

Commit:
Remove session field from config

Commit:
Remove session field from config

Commit:
Update README.md

Changes added to point to private documentation at dev-aura-comp-01. Badge removed as no status info can be provided by the local server for the time being

kawaits reviewed (changes requested):
Please update **documentation**:
 - session now is part of the model config:
        - architecture.rst#training
        - architecture.rst#session
        - architecture.rst#system configuration
        - examples/howto-model-handler.rst
 - fleximodel:
        - architecture.rst#models 
        - examples/howto-model-handler.rst
        - samples-> from training to dataset
           - explain implications for validation samples (no longer handled by the trainer)
        - explain new param "seed" and "shuffle" options
        - explain pros and cons of "duck-typing"
        - document new utils
        - document input_shape now part of model specific parameters
 - trackers:
        - complete current info at tracking.rst
 - platform, system and HW metrics CPU/GPU:
        - extend execution-results.rst metrics section
 - keras metrics:
        - extend execution-results.rst metrics section
 - preprocessors:
        - extend architecture.rst#datasets
 - kaggle datasets:
        - original documentation included info about 
           - how to proceed with kaggle datasets in terms of credentials
           - which datasets were supported
        - please recover this info and extend  datasets.rst with any additional recommendation (ie: when data loading fails due to a credential error )
 - nvidia-docker:
        - extend installing-gymnos.rst with info about how to get nvidia-docker, install it and how to get it up and running

kawaits commented on src/experiments/mnist.json:
```diff
@@ -5,21 +5,28 @@
-            {"type": "dense", "units": 64, "activation": "relu"},
-            {"type": "dense", "units": 10, "activation": "softmax"}
-        ]
+        "name": "keras",
```
1.- Experiment should be agnostic to comercial libraries.
2.- Name should be descriptive of the model not the library.  name:" keras" seems not to provide much info
Please find alternative. We can discuss it later

kawaits commented on src/experiments/tiny_imagenet.json:
```diff
@@ -5,23 +5,30 @@
-            "optimizer": "adam",
-            "loss": "categorical_crossentropy",
-            "metrics": ["acc"]
+        "name": "keras",
```
Idem

kawaits commented on src/experiments/unusual_data_usage_test.json:
```diff
@@ -11,16 +11,10 @@
-            "fit": 0.20,
-            "val": 0.20,
-            "test": 0.20
+            "train": 0.8,
```
I do not understand this: 
I thought from now on, samples were part of the "dataset" section not "training"

r-salas replied:
That's how it is.

kawaits commented on src/lib/models/keras.py:
```diff
@@ -0,0 +1,67 @@
+        self.model.compile(loss=loss, optimizer=optimizer, metrics=metrics)
+
+
+    def __build_compilation_from_config(self, loss, optimizer, metrics=None):
```
Why this method is not part of KerasMixin. What is the criteria?

r-salas replied:
You're right. I'm getting on it.

r-salas replied:
I'm reviewing the files and they are correctly defined. The keras mixin expects a `self.model` variable with a compiled keras model, it does not have a constructor. The keras model is the responsible for defining the model from the sequential and compilation parameters.

kawaits commented on src/lib/models/keras.py:
```diff
@@ -0,0 +1,67 @@
+        return loss, optimizer, metrics_funcs
+
+
+    def __build_sequential_from_config(self, sequential_config):
```
Why this method is not part of KerasMixin. What is the criteria?

r-salas replied:
You're right. I'm getting on it.

r-salas replied:
Idem

kawaits commented on src/lib/models/model.py:
```diff
@@ -4,86 +4,21 @@
+        - save(directory)
+        - restore(directory)
+
+    We use duck-typing but this class is defined with 2 purposes:
```
What's the advantage of having duck typing?
Duck typing is **more about convenience than type safety**. Source: Cui, 2015. 
It's been said that dynamic typing in general cuts down development time. There's less boilerplate code. Code is easier to hack. With static typing, compile-time checks slow down the development process. Static typing could be used at a later point to get better performance. **Others recommend duck typing only for prototyping and never for production.**

While it's true that compile-time checks are useful to catch potential problems early, duck typing enforces following coding conventions, documentation and test-driven methodologies.

Since duck typing isn't exactly a type system, it gives programmers flexibility. In Python, for example, common things are simpler to code. While static typed languages use interfaces, such interfaces may involve excessive refactoring of client code.

kawaits commented on src/lib/trainer.py:
```diff
@@ -36,16 +38,17 @@ class Trainer:
     """
 
-    def __init__(self, experiment, model, dataset, training, session=None, tracking=None):
+    def __init__(self, experiment, model, dataset, training, session, tracking):
```
Why we are still having session as an input for the trainer? It is used no where. 
No a model class responsability
Please modify accordingly

kawaits commented on src/bin/scripts/gymnosd.py:
```diff
@@ -45,10 +45,10 @@
         model=Model(**training_config["model"]),
         dataset=Dataset(cache_dir=cache_config["datasets"], **training_config["dataset"]),
         training=Training(**training_config["training"]),
+        experiment=Experiment(**training_config.get("experiment", {})),  # optional field
```
In which scenario experiment would not be required?

r-salas replied:
Right now, name and tags for experiment field are optional fields, so we could do something like this:
```json
{
   "experiment": {}
}
```
I don't think is a good idea to force to have this field if both attributes are optional.

Commit:
Added changes to Azure Pipelines to preserve privacy in docs 

* Update azure-pipelines.docker.yml for Azure Pipelines

* Update azure-pipelines.docker.yml for Azure Pipelines

* Update azure-pipelines.docker.yml for Azure Pipelines

Commit:
Better documentation for pipenv installation and execution

Commit:
Better documentation for pipenv installation and execution

Commit:
Upgrade Tensorflow version

Commit:
Upgrade Tensorflow version

Commit:
Fix documentation for pipenv shell directory

Commit:
Fix documentation for pipenv shell directory

Commit:
Fix sklearn regressormixin name

Commit:
Fix sklearn regressormixin name

Commit:
Fix lemmatization lenguage parameter

Commit:
Fix lemmatization lenguage parameter

Commit:
Fix kaggle decompressing

Commit:
Fix kaggle decompressing

Commit:
Change samples proportion

Commit:
Change samples proportion

Commit:
Format precision for samples ratio

Commit:
Format precision for samples ratio

Commit:
Fix variable typo

Commit:
Fix variable typo

Commit:
Summary of metrics written with logger not pprint

Commit:
Summary of metrics written with logger not pprint

Commit:
Decompress all files for kaggle datasets

Commit:
Decompress all files for kaggle datasets

Commit:
Move pipenv instructions to python block

Commit:
Move pipenv instructions to python block

Commit:
Fix numpy types saving JSON

Commit:
Fix numpy types saving JSON

Commit:
Add MTE model

Commit:
Add MTE model

Commit:
Fix validation for dogs vs cats model

Commit:
Fix validation for dogs vs cats model

Commit:
Remove unused module

Commit:
Remove unused module

Commit:
Close TF session before delete object

Commit:
Close TF session before delete object

Commit:
Convert sparse array to dense array

Commit:
Convert sparse array to dense array

Commit:
Add regression test and reorganize experiments directory

Commit:
Add regression test and reorganize experiments directory

Commit:
Update Gymnos by example documentation

Commit:
Merge branch 'devel-solving-issues' of https://github.com/Telefonica/gymnos into devel-solving-issues

Commit:
Remove transformers from Gymnos by Example

Commit:
Add sphinx package

Commit:
Core arguments for trainer in run method

Commit:
Better name for training config variable

Commit:
Add sphinx theme to dev packages

Commit:
Add Sphinx autodoc

Commit:
Change MTE name and add MTE documentation

Commit:
Update datasets documentation

Commit:
Update models documentation

Commit:
Fix typo

Commit:
Fix sphinx warnings

Commit:
Temporary solution for sphinx bug (sphinx issue 5995)

Commit:
Hide inherited members for datasets docs

Commit:
Remove handler name from filename

Commit:
Remove old preprocessor

Commit:
Add trackers documentation

Commit:
Add preprocessors documentation

Commit:
Add core and training config documentation

Commit:
Make explicit one-hot encoding option

Commit:
Update documentation index

Commit:
Temporary directory for regression test outputs

Commit:
Fix regression counter variable

Commit:
Update datasets for mixins

Commit:
Update documentation for model

Commit:
Better documentation for trackers

Commit:
Better documentation for preprocessors

Commit:
Add members to trackers docs

Commit:
Datasets with mixins

Commit:
Better models documentation

kawaits commented on Pipfile.lock:
```diff
@@ -43,17 +43,17 @@
         },
         "boto3": {
```
¿AWS?

r-salas replied:
MLFlow dependency.

kawaits commented on Pipfile.lock:
```diff
@@ -43,17 +43,17 @@
-            "version": "==1.9.131"
+            "version": "==1.9.134"
         },
         "botocore": {
```
¿AWS?

r-salas replied:
MLFlow dependency

kawaits commented on docs/source/trackers.rst:
```diff
@@ -0,0 +1,22 @@
+###############################
```
Please combine with  https://github.com/Telefonica/gymnos/blob/devel/docs/source/tracking.rst

Commit:
Add missing mock imports for autodoc

Commit:
Save and restore for preprocessors

Commit:
Save all training artifacts to directory

Commit:
Specify that path is a directory for model save and restore

Commit:
Better documentation for notes in Gymnos by Example

Commit:
Add mlflow to mock imports

Commit:
Add Kaggle credentials documentation

Commit:
Fix sphinx warning underline too short

Commit:
Full methods documentation for datasets

Commit:
Add missing mock imports

Commit:
Add missing mock imports for docs

Commit:
Show compression info for dataset mixins

Commit:
Fix statsmodels pickle bug (statsmodels issue 4772)

Commit:
Add two-step cache

Commit:
Add dill mock import

Commit:
Better logging for test metrics

Commit:
Add reference to original datasets for models

Commit:
Mean value for test metrics

Commit:
Rename trainer run method

Commit:
Auto-download spacy models

Commit:
Float format logging split dataset

Commit:
Remove redundant cometml method

Commit:
Add documentation when train and test samples are set to None

Commit:
Documentation for decompressor

Commit:
Add services documentation

Commit:
Add logger documentation

Commit:
Replace count -> len

Commit:
Add utils documentation

Commit:
Log if cache exists

Commit:
Handle logging to file when exception

Commit:
Remove redundant documentation

Commit:
Update gymnos in depth documentation

Commit:
Update documentation methods to show for dataset

Commit:
Better download cache handling

Commit:
Update tree documentation execution results

Commit:
Update comet.ml in depth documentation

Commit:
Update in depth execution results model and pipeline

Commit:
Reorganize docs for in depth & by example directory

Commit:
Add tensorflow config session builder from dict

r-salas commented:
This branch can be merged. 
New features:
- Regression test (#46)
- Updated and extended documentation (#16, #55, #66)
- HW details for metrics and logging (#64)
- Tensorflow session builder (#64)
- Preprocessors short logging (#67)
- Metrics short logging (#68)
- Remove Docker -c option (#69)
- Model flexibility (#73)
- Dataset flexibility
- Fix M+ API MTE (#74)
- Better handling Kaggle credentials (#75)
- Unify preprocessors and transformers (#81)
- Closes #82 (outdated)
- Updated TensorFlow version (#88)
- Add data usage model to regression test (#54)

Commit:
Instructions install docker different platforms

Commit:
Instructions install nvidia-docker

Commit:
Fix sphinx warning space between lines

Commit:
Remove old documentation images

Commit:
Update diagram documentation datasets
